### PR TITLE
Use valueWithPullback directly rather than the special curry version.

### PR DESCRIPTION
**Pull request by pschuh**

### CODE REVIEW

File: Sources/DeepLearning/Layer.swift [hunk 1]
```diff
@@ -36,12 +36,16 @@ public protocol Layer: Differentiable & KeyPathIterable
     func applied(to input: Input) -> Output
 }
 
-public extension Layer {
+public extension Layer
+    where TangentVector : AdditiveArithmetic, CotangentVector : AdditiveArithmetic,
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: To make this more error-prone, let's add these constraints to all differential operators defined in [AutoDiff.swift](https://github.com/apple/swift/blob/tensorflow/stdlib/public/core/AutoDiff.swift) and add FIXME's explaining that they are workarounds for [SR-9595](https://bugs.swift.org/browse/SR-9595).

File: Sources/DeepLearning/Layer.swift [hunk 2]
```diff
@@ -36,12 +36,16 @@ public protocol Layer: Differentiable & KeyPathIterable
...
+          Input.TangentVector : AdditiveArithmetic, Input.CotangentVector : AdditiveArithmetic {
     func valueWithPullback(at input: Input)
         -> (output: Output,
             pullback: (Output.CotangentVector)
                 -> (layerGradient: CotangentVector, inputGradient: Input.CotangentVector)) {
-        let (out, pullback) = _valueWithPullback(at: self, input, in: Self.applied(to:))
+        let (out, pullback) = valueWithPullback(at: input) { layer, input in
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: Just return it directly.

File: Sources/DeepLearning/Layer.swift (hunk 1 above)
Comment (dan-zheng) [reply Sources/DeepLearning/Layer.swift]: Adding constraints to differential operators is a very good idea.

File: Sources/DeepLearning/Layer.swift (hunk 2 above)
Comment (pschuh) [reply Sources/DeepLearning/Layer.swift]: Please note that the returned pullback is a labeled tuple: (layerGradient: CotangentVector, inputGradient: Input.CotangentVector)

Comment (rxwei) [reply Sources/DeepLearning/Layer.swift]: I see. I'd expect tuple coercion to work implicitly, but apparently my memory is from years ago.

**Pull request merged**